index.js

```diff
@@ -2,149 +2,75 @@
-            }
-          })
-        } else {
-          cb(null, {
-            // only accept script files
-            skip: !(stat.isFile() && scriptPattern.test(file)),
-            file: toLoad
-          })
-        }
-      })
-    }, (err, files) => {
-      if (err) {
-        next(err)
-        return
-      }
+function loadAllPlugins (fastify, opts) {
+  const defaultPluginOptions = opts.options
 
+  return loadPluginFiles(opts)
+    .then((files) => {
       const stats = [].concat(...files)
 
       const allPlugins = {}
-
+      const promises = []
       for (let i = 0; i < stats.length; i++) {
         const { skip, file, opts } = stats[i]
 
         if (skip) {
           continue
         }
 
-        try {
-          const plugin = require(file)
-          const pluginOptions = Object.assign({}, defaultPluginOptions)
-          const pluginMeta = plugin[Symbol.for('plugin-meta')] || {}
-          const pluginName = pluginMeta.name || file
+        const promise = importPlugin(file)
+          .then((plugin) => {
+            const pluginOptions = Object.assign({}, defaultPluginOptions)
+            const pluginMeta = plugin[Symbol.for('plugin-meta')] || {}
+            const pluginName = pluginMeta.name || file
 
-          if (opts && !plugin.autoPrefix) {
-            plugin.autoPrefix = opts.prefix
-          }
+            if (opts && !plugin.autoPrefix) {
+              plugin.autoPrefix = opts.prefix
+            }
 
-          if (plugin.autoload === false) {
-            continue
-          }
+            if (plugin.autoload === false) {
+              return null
+            }
 
-          if (plugin.autoPrefix) {
-            const prefix = pluginOptions.prefix || ''
-            pluginOptions.prefix = prefix + plugin.autoPrefix
-          }
+            if (plugin.autoPrefix) {
+              const prefix = pluginOptions.prefix || ''
+              pluginOptions.prefix = prefix + plugin.autoPrefix
+            }
 
-          if (plugin.prefixOverride !== undefined) {
-            pluginOptions.prefix = plugin.prefixOverride
-          }
+            if (plugin.prefixOverride !== undefined) {
+              pluginOptions.prefix = plugin.prefixOverride
+            }
 
-          if (allPlugins[pluginName]) {
-            throw new Error(`Duplicate plugin: ${pluginName}`)
-          }
+            if (allPlugins[pluginName]) {
+              throw new Error(`Duplicate plugin: ${pluginName}`)
+            }
 
-          allPlugins[pluginName] = {
-            plugin,
-            name: pluginName,
-            dependencies: pluginMeta.dependencies,
-            options: pluginOptions
-          }
-        } catch (err) {
-          next(enrichError(err))
-          return
-        }
-      }
+            allPlugins[pluginName] = {
+              plugin,
+              name: pluginName,
+              dependencies: pluginMeta.dependencies,
+              options: pluginOptions
+            }
+          })
```

Comment:
Would you mind moving the above in its own function?

Reply:
I moved it to a function.